### DISCUSSION TITLE
model.track() Speed Mismatch

### DISCUSSION THREAD

**Discussion by samhouinfinitekinetic** [ultralytics/ultralytics#21504] [Q&A] [answered]
Hi all,
I'm testing real-time tracking using a YOLO-Seg model (yolo11s-seg.pt) on a 30 fps video. I read frames with OpenCV and run model.track() on each one.

Reading frames and cropping takes ~4 ms (≈250 it/s in tqdm), but once I add tracking, latency jumps to ~40 ms per frame (≈25 it/s), making it unsuitable for real-time tracking.

However, when I enable verbose=True, the reported speed is only ~10 ms (preprocess + inference + postprocess). This suggests an unaccounted ~30 ms of overhead, perhaps from the tracker?

If so, why isn’t this overhead reflected in the model.track(..., persist=True, verbose=True) output? If the tracker alone adds ~30 ms of latency, what would be the proper way to achieve real-time tracking?

Here’s a minimal example:
```python
import time
import numpy as np
from tqdm import tqdm
from ultralytics import YOLO

model = YOLO('yolo11s-seg.pt')
frame = np.random.randint(low=0, high=256, size=(1296, 2317, 3), dtype=np.uint8)

with tqdm(total=1000) as pbar:
    for _ in range(1000):
        res = model.track(frame, persist=True, verbose=False)
        pbar.update(1)
        time.sleep(0.004)
```
This code shows 25 it/s on my machine, but if you set `verbose=True` and comment out `pbar.update(1)`, the output is around 10ms:
```
0: 384x640 (no detections), 10.0ms
Speed: 1.5ms preprocess, 10.0ms inference, 0.4ms postprocess per image at shape (1, 3, 384, 640)
```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @samhouinfinitekinetic, thank you for sharing these detailed observations and for providing a minimal example! 🚀 This is an automated response—an Ultralytics engineer will review your question and assist you soon.

We recommend reviewing our [Docs](https://docs.ultralytics.com/) for helpful [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples, as many common performance and tracking questions are covered there.

If this is a 🐛 Bug Report, thank you for including a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/), as this helps us quickly investigate the issue.

For custom training or performance ❓ Questions, please include additional details such as your hardware specifications, OS, and Ultralytics version to assist with troubleshooting. Also, check our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/) for optimizing speed and performance.

Join the Ultralytics community for more support! For real-time chat, visit [Discord](https://discord.com/invite/ultralytics) 🎧. For in-depth technical discussions, see [Discourse](https://community.ultralytics.com/). Or join us on the [Subreddit](https://www.reddit.com/r/Ultralytics/) to share and learn with others.

## Upgrade

First, please ensure you are running the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of the following up-to-date, verified environments with all dependencies, including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM ([GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/))
- **Amazon** Deep Learning AMI ([AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/))
- **Docker Image** ([Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for your patience—our team will follow up to help clarify the tracking speed and reporting!

**Reply by glenn-jocher** [reply, marked as answer]
The discrepancy you're observing is expected behavior - the `verbose=True` output only reports the detection model timing (preprocess + inference + postprocess) and doesn't include the tracking algorithm overhead. The additional ~30ms comes from the tracking algorithm (BoT-SORT by default) which performs object association, trajectory prediction, and ID management between frames.

For better real-time performance, try switching to ByteTrack which is generally faster: `model.track(frame, persist=True, tracker='bytetrack.yaml')`. You can also optimize by using a smaller model like `yolo11n-seg.pt`, reducing input resolution, or enabling [Re-Identification features selectively](https://docs.ultralytics.com/modes/track/#enabling-re-identification-reid) based on your use case requirements.

**Reply by samhouinfinitekinetic** [reply]
Thanks Glenn, that makes sense. Still, from a user’s perspective, it would be clearer if verbose=True included tracker overhead. Otherwise, users might underestimate total latency and choose a larger model without realizing the huge tracker overhead.

**Reply by glenn-jocher** [reply]
That's valid feedback! The current design separates detection and tracking timing since they're distinct components - the detection model timing is hardware-dependent while tracker overhead is more algorithmic. However, you're right that a combined timing report would be more useful for users planning real-time deployments.

For now, you can benchmark the full pipeline yourself as you're doing, or consider logging both components separately if you need detailed profiling. This kind of user experience improvement could definitely be worth considering for future updates.

**Reply by samhouinfinitekinetic** [reply]
Hi Glenn I have 2 follow up question, 
1. I followed your suggestion and resized the input frame from (1296, 2316) to (429, 768), keeping the aspect ratio with the longer side set to 768, which is the model size during training. I indeed saw a 2x speedup. However, I’m a bit curious: I thought model.track() internally resizes to imgsz=768 anyway? Why does pre-resizing the frame boosts performance?
2. Regarding BGR vs RGB: I trained the model using the simple one-liner: `model.train(data='data.yaml', imgsz=768, ...)`. During inference, I loop through video frames using OpenCV and run `model.track(source=frame, ...)` As OpenCV reads frames in BGR, do I need to convert the frame to RGB before passing it to model.track()? Or does Ultralytics handle the color format internally?
Thanks for your help!

**Reply by mwasela** [reply]
I've also been having issues with realtime rtsp streaming, during multiple objects present in a single frame, the video kind of glitches making the tracker loose its internal ID assignment and reassigns a new ID to same object, which is really bad for consistency tracking dependent on internal ID's.
 What could be the problem here. Using Yolov8s. When I load a prerecorded video, it performs way better than realtime actually

**Reply by glenn-jocher** [reply]
@samhouinfinitekinetic Great questions! 

1. Pre-resizing improves performance because while the model internally resizes to imgsz=768, starting with a smaller input reduces the computational cost of that resize operation itself. The preprocessing overhead scales with input size, so (429, 768) → 768 is much faster to process than (1296, 2316) → 768.

2. You don't need to convert BGR to RGB - Ultralytics automatically handles the BGR→RGB conversion when processing OpenCV frames internally.

@mwasela The ID reassignment during RTSP streaming is typically caused by inconsistent frame timing, network latency, or dropped frames that disrupt the tracker's temporal consistency. Try using `persist=True`, reducing the input resolution, or switching to ByteTrack which handles occlusions better: `model.track(source, tracker='bytetrack.yaml')`. RTSP streams are inherently less stable than recorded videos due to network variability.

**Reply by samhouinfinitekinetic** [reply]
> 1. Pre-resizing improves performance because while the model internally resizes to imgsz=768, starting with a smaller input reduces the computational cost of that resize operation itself. The preprocessing overhead scales with input size, so (429, 768) → 768 is much faster to process than (1296, 2316) → 768.

Thanks for your response!
However, I have a slightly different perspective regarding the idea of resizing before calling model.track. Here's a simple reproducible example:
```
from ultralytics import YOLO
import numpy as np
import cv2
from tqdm import tqdm

model = YOLO("models/yolo11n.pt")
for i in tqdm(range(1000)):
    img = np.random.randint(0, 255, (2560, 1440, 3), dtype=np.uint8)
    img = cv2.resize(img, (640, 360), interpolation=cv2.INTER_LINEAR)
    results = model.track(img, verbose=False, persist=True)[0]
```
As you can see, I resize the frame before passing it to model.track, and the speed is faster than if I pass in the original (2560, 1440) frame directly.
This improvement is not because the model internally resizes to imgsz=640. The model uses the LetterBox transform, which itself relies on cv2.resize to fit the original image (2560, 1440) into the target size (640, 360). So resizing externally vs. internally should be same.
The real reason for the speed difference appears to be the `apply_sparseoptflow` method in the tracking pipeline: https://github.com/ultralytics/ultralytics/blob/7dd9ec6c80dc0514e067566adca6d51eb9b2774d/ultralytics/trackers/utils/gmc.py#L300
In this function, every input frame is downscaled by a factor of `self.downscale`, which is always set to 2. So a large input like (2560, 1440) becomes (1280, 720) before sparse optical flow operations like `cv2.goodFeaturesToTrack`, `cv2.calcOpticalFlowPyrLK`, and frame copy like `frame.copy()`.
The cost of these functions scales with image size. On the other hand, if I resize the frame to (640, 360) before `model.track`, then the internal downscaling in `apply_sparseoptflow` brings it down to (320, 180), which dramatically reduces the cost of following computations.
It might be worth reconsidering whether `self.downscale` should always be fixed at 2. If the input frame is already small, additional downscaling may not be necessary or helpful.
To sum up:  
- If your camera is static, set `gmc_method` to `None` in `botsort.yaml`. This disables global motion compensation, allowing BoT-SORT to run at a speed comparable to ByteTrack. In this case, there's no need to resize your frame before calling `model.track`.
- If your camera is moving, it's best to manually tune the `downscale` parameter in `apply_sparseoptflow` for better performance. Avoid downscaling your frame before calling `model.track`, as it will be downscaled again by a factor of 2 internally.
- **Perhaps a `downscale` parameter could be added to `botsort.yaml` so users can tune it according to their frame size.**
- **Or perhaps there could be an additional parameter in `model.track`, such as `static_camera=True`, which would automatically disable gmc**

**Comment by Y-T-G** [comment]
You can also disable `gmc_method` for `botsort` to speedup tracking with BoTSORT.

https://docs.ultralytics.com/modes/track/#tracker-arguments
